### PR TITLE
[#12782] Add support for healthcare barcodes

### PR DESCRIPTION
**Pull request by stefanahman**
- [x] Improve README.md

### CODE REVIEW

File: lib/gs1/ai.rb [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+    FUNC_STAT = '7021'.freeze
+
+    # Revision status N4+X..20 (FNC1)
+    REV_STAT = '7022'.freeze
+
+    # Global Individual Asset Identifier (GIAI) of an assembly
+    GIAI_ASSEMBLY = '7023'.freeze
+
+    # Number of processor with ISO Country Code N4+N3+X..27 (FNC1)
+    PROCESSOR = '703s'.freeze
```
Comment (loveapoex) [inline lib/gs1/ai.rb]: All other are pure numbers, is this `s` intended?

File: spec/gs1/record_spec.rb [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+      expect(described_class).to have_received(:ai)
+    end
+  end
+
+  describe '#to_s' do
+    subject { record.to_s }
+
+    before { allow(data).to receive(:to_s) }
+
+    it 'calls .ai' do
```
Comment (loveapoex) [inline spec/gs1/record_spec.rb]: This example does not match its expectation

File: lib/gs1/check_digit.rb [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  # only the length.
+  #
+  class CheckDigit
+    VALID_LENGTHS = [7, 11, 12, 13, 17].freeze
+
+    class InvalidError < StandardError; end
+
+    def initialize(sequence)
+      @original_sequence = sequence
+      @sequence = sequence.split('').reverse
```
Comment (loveapoex) [inline lib/gs1/check_digit.rb]: Prefer `#chars` over `#split('')`?

File: lib/gs1/check_digit.rb [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    end
+
+    def calculate_digit
+      validate!
+
+      sub_from_nearest_higher_ten.to_s
+    end
+
+    def calculate_sequence_with_digit
+      @original_sequence + calculate_digit
```
Comment (loveapoex) [inline lib/gs1/check_digit.rb]: Prefer private `attr_reader`?

File: lib/gs1/check_digit.rb [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    end
+
+    def validate!
+      raise InvalidError, 'Invalid length' unless VALID_LENGTHS.include?(@sequence.size)
+    end
+
+    private
+
+    def multiplied_sequence
+      @multiplied_sequence ||= @sequence.each_with_index.map do |digit, idx|
```
Comment (loveapoex) [inline lib/gs1/check_digit.rb]: Prefer private `attr_reader`?

File: lib/gs1/batch.rb [hunk 6]
```diff
@@ -0,0 +1,35 @@
+module GS1
+  # Batch or lot number: AI (10)
+  #
+  # The GS1 Application Identifier (10) indicates that the GS1 Application Identifier data field contains a
+  # batch or lot number. The batch or lot number associates an item with information the manufacturer
+  # considers relevant for traceability of the trade item to which the element string is applied. The data may
+  # refer to the trade item itself or to items contained. The number may be, for example, a production lot
+  # number, a shift number, a machine number, a time, or an internal production code. The data is
+  # alphanumeric and may include all characters contained in figure 7.11-1.
```
Comment (loveapoex) [inline lib/gs1/batch.rb]: Where is figure 7.11-1?

File: lib/gs1/content.rb [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+  #
+  # The GTIN of the trade items contained is the GTIN of the highest level of trade item contained in the
+  # logistic unit.
+  #
+  # Note: This element string SHALL be used only on a logistic unit if:
+  #
+  # - the logistic unit is not itself a trade item; and
+  # - all trade items that are contained at the highest level have the same GTIN
+  #
+  # The check digit is explained in section 7.9. Its verification, which must be carried out in the
```
Comment (loveapoex) [inline lib/gs1/content.rb]: Where is this section?

File: lib/gs1/extensions/check_digit_validation.rb [hunk 8]
```diff
@@ -0,0 +1,17 @@
+module GS1
+  module Extensions
+    # Ensures correct check digit validation.
+    #
+    module CheckDigitValidation
+      def valid_check_digit?
+        CheckDigit.with_sequence(data[0..-2]) == data
```
Comment (loveapoex) [inline lib/gs1/extensions/check_digit_validation.rb]: Contains false positives FYI. Maybe it is unavoidable.

File: lib/gs1/extensions/gtin.rb [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+  module Extensions
+    # Extension for a GS1 GTIN. Ensures correct formating and validation.
+    #
+    module GTIN
+      def self.included(base)
+        base.extend         LengthValidation::ClassMethods
+        base.send :include, LengthValidation::InstanceMethods
+        base.send :include, CheckDigitValidation
+
+        base.valid_lengths [8, 12, 13, 14].freeze
```
Comment (loveapoex) [inline lib/gs1/extensions/gtin.rb]: These lengths does not match `VALID_LENGTHS`  in `CheckDigitValidation`, is it intended?

File: lib/gs1/extensions/gtin.rb [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+
+        base.lengths.each do |length|
+          define_method "to_gtin_#{length}" do
+            data.to_s.rjust(length, '0')
+          end
+        end
+      end
+
+      def to_s
+        to_gtin_14
```
Comment (loveapoex) [inline lib/gs1/extensions/gtin.rb]: Why was `to_gtin_14` chosen out of the four methods?

File: lib/gs1/batch.rb [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+  # processed together with the GTIN of the trade item to which it relates. When indicating this element
+  # string in the non-HRI text section of a barcode label, the following data title SHOULD be used (see
+  # also section 3.2): BATCH/LOT
+  #
+  # Source: https://www.gs1.org/sites/default/files/docs/barcodes/GS1_General_Specifications.pdf
+  #
+  class Batch < Record
+    include Extensions::String
+
+    valid_lengths 1..20
```
Comment (loveapoex) [inline lib/gs1/batch.rb]: Can you point me to the implementation of this method?

File: lib/gs1/check_digit.rb [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+        idx.even? ? digit.to_i * 3 : digit.to_i * 1
+      end
+    end
+
+    def sub_from_nearest_higher_ten
+      (10 - (multiplied_sequence.inject(:+) % 10)) % 10
```
[the remark this answers is not in the document]
Comment (stefanahman) [reply lib/gs1/check_digit.rb]: No, if `multiplied_sequence.inject(:+)` sums up to `60`, we get `(10 - 0) % 10`

File: spec/gs1/check_digit_spec.rb [hunk 13]
```diff
@@ -0,0 +1,70 @@
+RSpec.describe GS1::CheckDigit do
+  let(:check_digit) { described_class.new(sequence) }
+  let(:sequence) { '629104150021' }
```
[the remark this answers is not in the document]
Comment (stefanahman) [reply spec/gs1/check_digit_spec.rb]: This is the example from: http://www.gs1.org/how-calculate-check-digit-manually

File: lib/gs1/content.rb (hunk 7 above)
Comment (stefanahman) [reply lib/gs1/content.rb]: In https://www.gs1.org/sites/default/files/docs/barcodes/GS1_General_Specifications.pdf as referenced in the bottom of the description.

File: lib/gs1/batch.rb (hunk 11 above)
Comment (stefanahman) [reply lib/gs1/batch.rb]: https://www.gs1.org/sites/default/files/docs/barcodes/GS1_General_Specifications.pdf in figure 3.4.1-1

File: lib/gs1/content.rb (hunk 7 above)
Comment (stefanahman) [reply lib/gs1/content.rb]: please review the newest changes

File: lib/gs1/validations/length_validation.rb [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+      def valid_length?(options)
+        normalize_allowed(options[:allowed]).include?(data&.size)
+      end
+
+      private
+
+      def normalize_allowed(allowed)
+        case allowed
+        when Range then allowed.to_a
+        when Array then allowed.flatten
```
Comment (loveapoex) [inline lib/gs1/validations/length_validation.rb]: When would you need this?

File: lib/gs1/check_digit.rb (hunk 12 above)
[the remark this answers is not in the document]
Comment (loveapoex) [reply lib/gs1/check_digit.rb]: No, it says `10     -     (multiplied_sequence.inject(:+) % 10)` which means 10 - the last digit.

File: spec/gs1/check_digit_spec.rb (hunk 13 above)
[the remark this answers is not in the document]
Comment (loveapoex) [reply spec/gs1/check_digit_spec.rb]: I still think it's a valid point. If i remove the reverse call the specs still passes which is bad behavior.

File: lib/gs1/batch.rb (hunk 6 above)
Comment (stefanahman) [reply lib/gs1/batch.rb]: https://www.gs1.org/sites/default/files/docs/barcodes/GS1_General_Specifications.pdf

File: lib/gs1/extensions/gtin.rb (hunk 10 above)
Comment (stefanahman) [reply lib/gs1/extensions/gtin.rb]: I've added a comment

File: lib/gs1/validations/length_validation.rb (hunk 14 above)
Comment (stefanahman) [reply lib/gs1/validations/length_validation.rb]: In case someone types `[[5]]`

File: lib/gs1/extensions/gtin.rb (hunk 9 above)
Comment (stefanahman) [reply lib/gs1/extensions/gtin.rb]: `CheckDigitCalculator` validates the input sequence without the check digit, thus `VALID_LENGTHS - 1`

File: lib/gs1/extensions/check_digit_validation.rb (hunk 8 above)
Comment (stefanahman) [reply lib/gs1/extensions/check_digit_validation.rb]: Related to your other comment?

File: spec/gs1/check_digit_spec.rb (hunk 13 above)
[the remark this answers is not in the document]
Comment (stefanahman) [reply spec/gs1/check_digit_spec.rb]: I agree

File: spec/gs1/check_digit_spec.rb [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+    context 'when invalid length' do
+      let(:sequence) { '123' }
+
+      it 'fails' do
+        expect { subject }.to raise_error(GS1::CheckDigit::InvalidError, 'Invalid length')
+      end
+    end
+  end
+
+  describe '#calculate_digit' do
```
[the remark this answers is not in the document]
Comment (stefanahman) [reply spec/gs1/check_digit_spec.rb]: What kind of calculations are legit?

File: lib/gs1/ai.rb (hunk 1 above)
Comment (stefanahman) [reply lib/gs1/ai.rb]: Removed!

File: lib/gs1/validations/length_validation.rb (hunk 14 above)
Comment (stefanahman) [reply lib/gs1/validations/length_validation.rb]: I can change, seems dumb when you say it :D

**Pull request merged**